File: 04-Data-Structures/code/typescript/src/doublyLinkedList.ts
import { DSError } from "./errors.js";

class DNode<T> {
  value: T | undefined;
  prev: DNode<T>;
  next: DNode<T>;
  constructor(value?: T) {
    this.value = value;
    this.prev = this;
    this.next = this;
  }
}

/** Doubly linked list with a circular sentinel node, giving O(1) operations at both ends. */
export class DoublyLinkedList<T> {
  private sentinel: DNode<T> = new DNode<T>();
  private length = 0;

  size(): number {
    return this.length;
  }

  pushFront(x: T): void {
    this.insertAfter(this.sentinel, x);
  }

  pushBack(x: T): void {
    this.insertAfter(this.sentinel.prev, x);
  }

  popFront(): T {
    if (this.length === 0) throw new DSError("empty", "popFront from empty list");
    return this.removeNode(this.sentinel.next);
  }

  popBack(): T {
    if (this.length === 0) throw new DSError("empty", "popBack from empty list");
    return this.removeNode(this.sentinel.prev);
  }

  private insertAfter(node: DNode<T>, x: T): void {
    const n = new DNode(x);
    n.prev = node;
    n.next = node.next;
    node.next.prev = n;
    node.next = n;
    this.length++;
  }

  private removeNode(node: DNode<T>): T {
    node.prev.next = node.next;
    node.next.prev = node.prev;
    this.length--;
    return node.value as T;
  }

  toList(): T[] {
    const out: T[] = [];
    let cur = this.sentinel.next;
    while (cur !== this.sentinel) {
      out.push(cur.value as T);
      cur = cur.next;
    }
    return out;
  }

  checkInvariants(): void {
    let count = 0;
    let cur = this.sentinel.next;
    while (cur !== this.sentinel) {
      count++;
      if (count > this.length + 1) throw new DSError("invalid", "cycle detected");
      if (cur.next.prev !== cur) throw new DSError("invalid", "broken back-link");
      cur = cur.next;
    }
    if (count !== this.length) throw new DSError("invalid", "length mismatch");
  }
}

File: 04-Data-Structures/code/python/seb_ds/doubly_linked_list.py
"""Doubly linked list with a circular sentinel node. Mirrors `typescript/src/doublyLinkedList.ts`."""

from __future__ import annotations

from typing import Generic, List, Optional, TypeVar

from .errors import DSError

T = TypeVar("T")


class _DNode(Generic[T]):
    __slots__ = ("value", "prev", "next")

    def __init__(self, value: Optional[T] = None) -> None:
        self.value = value
        self.prev: "_DNode[T]" = self
        self.next: "_DNode[T]" = self


class DoublyLinkedList(Generic[T]):
    def __init__(self) -> None:
        self._sentinel: _DNode[T] = _DNode()
        self._length = 0

    def size(self) -> int:
        return self._length

    def push_front(self, x: T) -> None:
        self._insert_after(self._sentinel, x)

    def push_back(self, x: T) -> None:
        self._insert_after(self._sentinel.prev, x)

    def pop_front(self) -> T:
        if self._length == 0:
            raise DSError("empty", "popFront from empty list")
        return self._remove_node(self._sentinel.next)

    def pop_back(self) -> T:
        if self._length == 0:
            raise DSError("empty", "popBack from empty list")
        return self._remove_node(self._sentinel.prev)

    def _insert_after(self, node: _DNode[T], x: T) -> None:
        n = _DNode(x)
        n.prev = node
        n.next = node.next
        node.next.prev = n
        node.next = n
        self._length += 1

    def _remove_node(self, node: _DNode[T]) -> T:
        node.prev.next = node.next
        node.next.prev = node.prev
        self._length -= 1
        return node.value  # type: ignore[return-value]

    def to_list(self) -> List[T]:
        out: List[T] = []
        cur = self._sentinel.next
        while cur is not self._sentinel:
            out.append(cur.value)  # type: ignore[arg-type]
            cur = cur.next
        return out

    def check_invariants(self) -> None:
        count = 0
        cur = self._sentinel.next
        while cur is not self._sentinel:
            count += 1
            if count > self._length + 1:
                raise DSError("invalid", "cycle detected")
            if cur.next.prev is not cur:
                raise DSError("invalid", "broken back-link")
            cur = cur.next
        if count != self._length:
            raise DSError("invalid", "length mismatch")
